BaragonService/src/main/java/com/hubspot/baragon/service/elb/ApplicationLoadBalancer.java

```diff
@@ -29,23 +62,401 @@ public ApplicationLoadBalancer(Optional<ElbConfiguration> configuration,
+          guaranteeHasAllSubnets(
+              agent.getEc2().getSubnetId().asSet(),
+              loadBalancer);
+        }
+      }
+    }
+  }
+
+  /**
+   *
+   * @param target Target to check
+   * @param targetGroup Group to check in
+   * @return if the given target is the last healthy target in the given target group
+   */
+  private boolean isLastHealthyInstance(TargetDescription target, TargetGroup targetGroup) {
+    DescribeTargetHealthRequest targetHealthRequest = new DescribeTargetHealthRequest()
+        .withTargetGroupArn(targetGroup.getTargetGroupArn());
+    List<TargetHealthDescription> targetHealthDescriptions = elbClient
+        .describeTargetHealth(targetHealthRequest)
+        .getTargetHealthDescriptions();
+
+    boolean instanceIsHealthy = false;
+    int healthyCount = 0;
+
+    for (TargetHealthDescription targetHealthDescription : targetHealthDescriptions) {
+      if (targetHealthDescription.getTargetHealth().getState()
+          .equals(TargetHealthStateEnum.Healthy.toString())) {
+        healthyCount += 1;
+        if (targetHealthDescription.getTarget().equals(target)) {
+          instanceIsHealthy = true;
+        }
+      }
+    }
+
+    return instanceIsHealthy && healthyCount == 1;
+  }
+
+  private void guaranteeHasAllSubnets(Collection<String> subnetIds, LoadBalancer loadBalancer) {
+    Collection<String> subnetsOnLoadBalancer = getSubnetsFromLoadBalancer(loadBalancer);
+    Collection<String> subnetsToAdd = new HashSet<>();
+    for (String subnetId : subnetIds) {
+      if (! subnetsOnLoadBalancer.contains(subnetId)) {
+        subnetsToAdd.add(subnetId);
+      }
+    }
+
+    try {
+      SetSubnetsRequest subnetsRequest = new SetSubnetsRequest()
+          .withLoadBalancerArn(loadBalancer.getLoadBalancerArn())
+          .withSubnets(subnetsToAdd);
+      elbClient.setSubnets(subnetsRequest);
+    } catch (AmazonClientException acexn) {
+      LOG.error("Could not attach subnets {} to load balancer {} due to error",
+          subnetsToAdd, loadBalancer.getLoadBalancerName(), acexn);
+      exceptionNotifier.notify(acexn, ImmutableMap.of(
+          "elb", loadBalancer.getLoadBalancerName(),
+          "subnets", subnetsToAdd.toString()));
+    }
+  }
+
+  /**
+   * After this method completes, every agent in baragonAgents should be associated with
+   * a target in the given target group.
+   *
+   * @param targetGroup group to register in
+   * @param baragonAgents agents to be registered
+   */
+  private void guaranteeHasAllTargets(TargetGroup targetGroup,
+                                      Collection<BaragonAgentMetadata> baragonAgents) {
+    Collection<TargetDescription> targetDescriptions = new HashSet<>();
+    for (BaragonAgentMetadata agent : baragonAgents) {
+      try {
+        if (agent.getEc2().getInstanceId().isPresent()) {
+          if (agentShouldRegisterInTargetGroup(agent.getEc2().getInstanceId().get(), targetGroup)) {
```

Comment:
Another case of being able to reuse data from a single api call. We know these agents are all in the same baragon group, and therefore the same target group. We an make the describe target health call before this method and reuse that data for each agent we check